sonar-db/src/main/java/org/sonar/core/timemachine/Periods.java

```diff
@@ -40,88 +47,107 @@ public Periods(Settings settings, I18n i18n) {
-      label = label("since_x", shortLabel, date);
-    } else {
-      throw new IllegalArgumentException("This mode is not supported : " + mode);
+    switch (mode) {
+      case TIMEMACHINE_MODE_DAYS:
+        return labelForDays(param, date, shortLabel);
+      case TIMEMACHINE_MODE_VERSION:
+        return labelForVersion(param, date, shortLabel);
+      case TIMEMACHINE_MODE_PREVIOUS_ANALYSIS:
+        return labelForPreviousAnalysis(date, shortLabel);
+      case TIMEMACHINE_MODE_PREVIOUS_VERSION:
+        return labelForPreviousVersion(param, date, shortLabel);
+      case TIMEMACHINE_MODE_DATE:
+        return label("since_x", shortLabel, date);
+      default:
+        throw new IllegalArgumentException("This mode is not supported : " + mode);
     }
-    return label;
   }
 
-  private String label(String key, boolean shortLabel, Object... parameters) {
-    String msgKey = key;
-    if (shortLabel) {
-      msgKey += ".short";
+  private String labelForDays(@Nullable String param, @Nullable String date, boolean shortLabel) {
+    if (date == null) {
+      return label("over_x_days", shortLabel, param);
     }
-    return i18n.message(getLocale(), msgKey, null, parameters);
+    return label("over_x_days_detailed", shortLabel, param, date);
   }
 
-  @CheckForNull
-  private static String convertDate(Date date) {
-    if (date != null) {
-      SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy MMM dd");
-      return dateFormat.format(date);
+  private String labelForVersion(@Nullable String param, @Nullable String date, boolean shortLabel) {
+    if (date == null) {
+      return label("since_version", shortLabel, param);
     }
-    return null;
+    return label("since_version_detailed", shortLabel, param, date);
   }
 
-  private static Locale getLocale() {
-    return Locale.ENGLISH;
+  private String labelForPreviousAnalysis(@Nullable String date, boolean shortLabel) {
+    if (date == null) {
+      return label("since_previous_analysis", shortLabel);
+    }
+    return label("since_previous_analysis_detailed", shortLabel, date);
+  }
+
+  private String labelForPreviousVersion(@Nullable String param, @Nullable String date, boolean shortLabel) {
+    if (param == null && date == null) {
+      return label("since_previous_version", shortLabel);
+    }
+    if (param == null) {
+      // Special case when no snapshot for previous version is found. The first analysis is then returned -> Display only the date.
+      return label("since_previous_version_with_only_date", shortLabel, date);
+    }
+    if (date == null) {
+      return label("since_previous_version_detailed", shortLabel, param);
+    }
+    return label("since_previous_version_detailed", shortLabel, param, date);
+  }
+
+  private String label(String key, boolean shortLabel, Object... parameters) {
+    String msgKey = key;
+    if (shortLabel) {
+      msgKey += ".short";
+    }
+    return i18n.message(getLocale(), msgKey, null, parameters);
```

Comment:
What about inlining `Local.ENGLISH`  instead of using a private method for a single call ?